src/prognostic_equations/edmfx_closures.jl

```diff
@@ -206,78 +228,187 @@ function mixing_length(
+    # N²/Pr_t denotes the gradient involved in buoyancy production.
+    # The factor below corresponds to that production term normalised by l.
+    a_pd = c_m * (2 * ᶜstrain_rate_norm - ᶜlinear_buoygrad / ᶜPr) * sqrt_tke_pos
+
+    # Dissipation is modelled as c_d · k^{3/2} / l.
+    # For the quadratic expression below, c_neg ≡ c_d · k^{3/2}.
+    c_neg = c_d * tke_pos * sqrt_tke_pos
+
+    l_TKE = FT(0)
+    # Solve for l_TKE in
+    #     a_pd · l_TKE − c_neg / l_TKE + ᶜtke_exch = 0
+    #  ⇒  a_pd · l_TKE² + ᶜtke_exch · l_TKE − c_neg = 0
+    # yielding
+    #     l_TKE = (−ᶜtke_exch ± √(ᶜtke_exch² + 4 a_pd c_neg)) / (2 a_pd).
+    if abs(a_pd) > eps_FT # If net of shear and buoyancy production (a_pd) is non-zero
+        discriminant = ᶜtke_exch * ᶜtke_exch + 4 * a_pd * c_neg
+        if discriminant >= FT(0) # Ensure real solution exists
+            # Select the physically admissible (positive) root for l_TKE.
+            # When a_pd > 0 (production exceeds dissipation) the root
+            #     (−ᶜtke_exch + √D) / (2 a_pd)
+            # is positive.  For a_pd < 0 the opposite root is required.
+            l_TKE_sol1 = (-(ᶜtke_exch) + sqrt(discriminant)) / (2 * a_pd)
+            # For a_pd < 0 (local destruction exceeds production) use
+            #     (−ᶜtke_exch − √D) / (2 a_pd).
+            if a_pd > FT(0)
+                l_TKE = l_TKE_sol1
+            else # a_pd < FT(0)
+                l_TKE = (-(ᶜtke_exch) - sqrt(discriminant)) / (2 * a_pd)
+            end
+            l_TKE = max(l_TKE, FT(0)) # Ensure it's non-negative
+        end
+    elseif abs(ᶜtke_exch) > eps_FT # If a_pd is zero, balance is between exchange and dissipation
+        # ᶜtke_exch = c_neg / l_TKE  => l_TKE = c_neg / ᶜtke_exch
+        # Ensure division is safe and result is positive
+        if ᶜtke_exch > eps_FT # Assuming positive exchange means TKE sink from env perspective
+            l_TKE = c_neg / ᶜtke_exch # if c_neg is positive, l_TKE is positive
+        elseif ᶜtke_exch < -eps_FT # Negative exchange means TKE source for env
+            # -|ᶜtke_exch| = c_neg / l_TKE. If c_neg > 0, this implies l_TKE < 0, which is unphysical.
+            # This case (a_pd=0, tke_exch < 0, c_neg > 0) implies TKE source and dissipation, no production.
+            # Dissipation = Source. So, c_d * k_sqrt_k / l = -tke_exch. l = c_d * k_sqrt_k / (-tke_exch)
+            l_TKE = c_neg / (-(ᶜtke_exch))
+        end
+        l_TKE = max(l_TKE, FT(0))
     end
-
-    # compute l_N - the effective static stability length scale.
-    N_eff = sqrt(max(ᶜlinear_buoygrad, 0))
-    if N_eff > 0.0
-        l_N = min(sqrt(max(c_b * ᶜtke, 0)) / N_eff, l_z)
-    else
-        l_N = l_z
+    # If a_pd = 0 and ᶜtke_exch = 0 (or c_neg = 0), l_TKE remains zero.
+
+    # --- l_N: Static-stability length scale (buoyancy limit), constrained by l_z ---
+    N_eff_sq = max(ᶜlinear_buoygrad, FT(0)) # Use N^2 only if stable (N^2 > 0)
```

Comment:
```suggestion
    N_eff_sq = max(ᶜlinear_buoygrad, eps_FT) # Use N^2 only if stable (N^2 > 0)
```
(To guard against division by zero later.)